components/datalogger/datalogger_utils.py

```diff
@@ -714,3 +721,112 @@ def ProcessDataloggerToCsv(a_server, a_site_id, a_headers, a_target_dir, a_datal
+                "finger_print": self.certificate.digest("sha1").decode('utf-8'),
+                "finger_print_algorithm": "SHA1",
+                "public_id": {
+                    "type":"RSA",
+                    "public_pem_b64": base64.b64encode(certificate_pem).decode('utf-8'),
+                }
+            },
+            "at": at or config.get_milliseconds_since_epoch(),
+            "ttl": 57600,
+        }
+
+class AssetContext():
+    def __init__(self, assets, lease_start = None):
+        self.assets = assets
+        self.uuid = str(uuid.uuid4())
+        self.lease_start = lease_start or config.get_milliseconds_since_epoch()
+        self.lease_ttl = 28800
+
+    def make_payload(self):
+        payload = {
+            "uuid": self.uuid,
+            "lease_start": self.lease_start,
+            "lease_ttl": self.lease_ttl,
+        }
+        document = json.dumps(payload)
+        document_b64 = base64.b64encode(document.encode('utf-8')).decode('utf-8')
+        def get_asset_signature(asset):
+            return {
+                "algorithm": "SHA256-RSA",
+                "asset_urn": asset.get_urn(),
+                "asset_uuid": asset.asset_uuid,
+                "signature_b64": base64.b64encode(openssl.sign(asset.key_pair, document, "sha256")).decode('utf-8'),
+            }
+        return {
+            "type": "sitelink::AssetContext",
+            "data": {
+                "uuid": self.uuid,
+                "lease_start": self.lease_start,
+                "lease_ttl": self.lease_ttl,
+                "signatures": [get_asset_signature(asset) for asset in self.assets],
+                "document_b64": document_b64,
+            },
+            "at": self.lease_start,
+            "ttl": self.lease_ttl,
+        }
+
+class DataLoggerAggregatorSocket():
+    def __init__(self, a_server_config, a_site_id, a_dc, a_domain, a_region):
+        self.site_id = a_site_id
+        self.domain = a_domain
+        self.region = a_region
+        self.ws = websocket.WebSocket(sslopt={"cert_reqs": ssl.CERT_NONE})
+        self.ws.settimeout(2)
+        dl_ws_url = "{}/data_logger/v1/publish_live/{}/{}/{}.{}".format(a_server_config.to_url(), a_dc, a_domain, a_site_id, a_region)
+        self.ws.connect(dl_ws_url)
+    def send(self, payload):
+        return self.ws.send(payload)
+    def receive(self):
+        return self.ws.recv()
+    def receive_ack(self):
+        return map(int, self.ws.recv().split("\n"))
+        
+
+def make_haul_mfk_replicate_payload(asset_context, rc_uuid, latitude, longitude, altitude, direction, at = None):
+    manifest = io.BytesIO()
+    manifest.write(struct.pack("<d", latitude))
+    manifest.write(struct.pack("<d", longitude))
+    manifest.write(struct.pack("<d", altitude))
+    manifest.write(struct.pack("<f", direction))
+    manifest.seek(0)
+    return {
+        "type": "mfk::Replicate",
+        "data": {
+            "ac_uuid": asset_context.uuid,
+            "beacon": True,
```

Comment:
beacon flag should be set according to timestamp

Reply:
Thanks, have implemented for low frequency beacon.